source/common/router/router.cc

```diff
@@ -491,6 +493,34 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::RequestHeaderMap& headers,
 
   transport_socket_options_ = Network::TransportSocketOptionsUtility::fromFilterState(
       *callbacks_->streamInfo().filterState());
+
+  if (callbacks_->getUpstreamSocketOptions()) {
```

Comment:
I'm not sure if its worth the complexity or not, but I think we can avoid an allocation in most cases:

Move this chunk under the `if(has_downstream_connection_with_redirect_records)` block below, and change `upstreamSocketOptions()` to `return (upstream_options_ != nullptr) ? upstream_options_ : callbacks_->getUpstreamSocketOptions();`

Reply:
fixed in 2843f25025763c74a05673d5c618d9f83ce72e1c